### PR TITLE
GH-35659: [Swift] Initial Swift IPC writer

### LINKED ISSUE #35659: [Swift] Add initial IPC writer support
### Describe the enhancement requested

Add initial implementation of IPC writer for swift impl of arrow.

### Component(s)

Swift

### PR DESCRIPTION
**Pull request by abandy**
### Rationale for this change
This adds the initial IPC writer for swift in order to write the arrow data to memory or to a file.

### What changes are included in this PR?

- IPC writer
- bug fixes found during writer testing

### Are these changes tested?

Writer unit test added.
* Closes: #35659

### CODE REVIEW

File: swift/Arrow/Tests/ArrowTests/IPCTests.swift [hunk 1]
```diff
@@ -53,28 +78,83 @@ final class IPCTests: XCTestCase {
...
+        let arrowReader = ArrowReader()
+        let fileRBs = try arrowReader.fromFile(fileURL)
+        checkBoolRecordBatch(fileRBs)
+        let arrowWriter = ArrowWriter()
+        //write data from file to a stream
+        let writeData = try arrowWriter.toStream(fileRBs[0].schema, batches: fileRBs)
+        //read stream back into recordbatches
+        checkBoolRecordBatch(try arrowReader.fromStream(writeData))
+        //write file record batches to another file
+        let outputUrl = currentDirectory().appendingPathComponent("../../testfilereader_bool.arrow")
```
Comment (kou) [inline swift/Arrow/Tests/ArrowTests/IPCTests.swift:97]:
```suggestion
        let outputUrl = currentDirectory().appendingPathComponent("../../testfilewriter_bool.arrow")
```

**Review by kou — APPROVED** [review]
+1

File: swift/Arrow/Sources/Arrow/ArrowTable.swift [hunk 2]
```diff
@@ -22,18 +22,58 @@ public class ChunkedArrayHolder {
...
+    public let getBufferDataSizes: () throws -> [Int]
     public init<T>(_ chunked: ChunkedArray<T>) {
         self.holder = chunked
         self.length = chunked.length
         self.type = chunked.type
         self.nullCount = chunked.nullCount
+        self.getBufferData = {() throws -> [Data] in 
+            var bufferData = [Data]()
+            var numBuffers = 2;
+            if(!isFixedPrimitive(try toFBTypeEnum(chunked.type))) {
```
[the remark this answers is not in the document]
Comment (abandy) [reply swift/Arrow/Sources/Arrow/ArrowTable.swift:35]: Good catch, will update.

File: swift/Arrow/Sources/Arrow/ArrowTable.swift [hunk 3]
```diff
@@ -22,18 +22,58 @@ public class ChunkedArrayHolder {
...
+                }
+            }
+
+            return bufferData;
+        }
+        
+        self.getBufferDataSizes = {() throws -> [Int] in
+            var bufferDataSizes = [Int]()
+            var numBuffers = 2;
+            if(!isFixedPrimitive(try toFBTypeEnum(chunked.type))) {
```
[the remark this answers is not in the document]
Comment (abandy) [reply swift/Arrow/Sources/Arrow/ArrowTable.swift:55]: Will update.

File: swift/Arrow/Sources/Arrow/ArrowReader.swift [hunk 4]
```diff
@@ -36,11 +36,11 @@ public class ArrowReader {
     }
 
     private func loadSchema(_ schema: org_apache_arrow_flatbuf_Schema) throws -> ArrowSchema {
-        let builder = ArrowSchema.Builder()
+        let builder: ArrowSchema.Builder = ArrowSchema.Builder()
```
[the remark this answers is not in the document]
Comment (abandy) [reply swift/Arrow/Sources/Arrow/ArrowReader.swift:39]: This was added by xcode.  I will remove.

File: swift/Arrow/Sources/Arrow/ArrowReader.swift [hunk 5]
```diff
@@ -36,11 +36,11 @@ public class ArrowReader {
...
         for index in 0 ..< schema.fieldsCount {
             let field = schema.fields(at: index)!
-            let arrowField = ArrowField(field.name!, type: findArrowType(field), isNullable: field.nullable)
-            builder.addField(arrowField)
+            let arrowField: ArrowField = ArrowField(field.name!, type: findArrowType(field), isNullable: field.nullable)
+            let _ = builder.addField(arrowField)
```
[the remark this answers is not in the document]
Comment (abandy) [reply swift/Arrow/Sources/Arrow/ArrowReader.swift:43]: The return value is just the Builder so we can chain the calls for other use cases.

File: swift/Arrow/Sources/Arrow/ArrowWriter.swift [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+            self.handle.write(data)
+            self.current_size += data.count
+        }
+    }
+
+    private func writeField(_ fbb: inout FlatBufferBuilder, field: ArrowField) throws -> Offset {
+        let nameOffset = fbb.create(string: field.name)
+        let fieldTypeOfffset = try toFBType(&fbb, infoType: field.type)
+        let startOffset = org_apache_arrow_flatbuf_Field.startField(&fbb)
+        org_apache_arrow_flatbuf_Field.add(name: nameOffset , &fbb)
```
[the remark this answers is not in the document]
Comment (abandy) [reply swift/Arrow/Sources/Arrow/ArrowWriter.swift:60]: Will update.

File: swift/Arrow/Tests/ArrowTests/IPCTests.swift (hunk 1 above)
Comment (abandy) [reply swift/Arrow/Tests/ArrowTests/IPCTests.swift:97]: Will update.

**Comment by abandy** [comment]
Hi @kou, thank you!  I have updated accordingly and will merge after the changes based on your comments have passed checks.

File: swift/Arrow/Sources/Arrow/ArrowReader.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (kou) [reply swift/Arrow/Sources/Arrow/ArrowReader.swift:43]: I see.

**Pull request merged**